### PR TITLE
Updates for latest warnings and deprecations.

### PR DESCRIPTION
**Pull request by milessabin**

### CODE REVIEW

File: core/src/main/scala/shapeless/generic.scala [hunk 1]
```diff
@@ -743,6 +743,22 @@ trait CaseClassMacros extends ReprTypes with CaseClassMacrosVersionSpecifics {
...
 
+  def numNonCaseParamLists(tpe: Type): Int = {
+    val companion = patchedCompanionSymbolOf(tpe.typeSymbol).typeSignature
+    val apply = companion.member(TermName("apply"))
+    if (apply.isMethod && !isNonGeneric(apply) && isAccessible(companion, apply)) {
+      val paramLists = apply.typeSignatureIn(companion).paramLists
+      val numParamLists = paramLists.length
+      if (numParamLists <= 1) 0
+      else {
+        if (paramLists.last.headOption.map(_.isImplicit).getOrElse(false))
```
Comment (joroKr21) [inline core/src/main/scala/shapeless/generic.scala:754]:
```suggestion
        if (paramLists.last.headOption.exists(_.isImplicit))
```

File: build.sbt [hunk 2]
```diff
@@ -162,6 +168,7 @@ lazy val scratch = crossProject(JSPlatform, JVMPlatform, NativePlatform).crossTy
   .jsSettings(commonJsSettings:_*)
   .jvmSettings(commonJvmSettings:_*)
   .nativeSettings(commonNativeSettings:_*)
+  .jvmSettings(scalacOptions += "-Xprint:typer")
```
[the remark this answers is not in the document]
Comment (milessabin) [reply build.sbt:171]: Oops!

File: examples/src/main/scala/shapeless/examples/linearalgebra.scala [hunk 3]
```diff
@@ -52,8 +52,10 @@ object LinearAlgebraExamples extends App {
         mapper : Mapper.Aux[sum.type, ZLN, LN]) : PN => VectorOps[N, PN] =
           (p : PN) =>
             new VectorOps[N, PN](p) {
-              def +(other : Self) : Self =
-                newtype(gen.from((gen.to(p) :: gen.to(other.tupled) :: HNil).transpose.map(sum)))
+              def +(other : Self) : Self = {
+                val other0 = newtype.newtypeOps(other)(pointOpsN)
+                newtype(gen.from((gen.to(p) :: gen.to(other0.tupled) :: HNil).transpose.map(sum)))
```
[the remark this answers is not in the document]
Comment (milessabin) [reply examples/src/main/scala/shapeless/examples/linearalgebra.scala:57]: It's a workaround for a linter warning. There's possibly a better fix, but I don't think it's worth spending too much time on these (now ancient) examples.

File: core/src/main/scala/shapeless/generic.scala (hunk 1 above)
Comment (milessabin) [reply core/src/main/scala/shapeless/generic.scala:754]: :+1:

**Pull request merged**